README.md
<p align="center">
    <img alt="logo" src="https://cdn.jsdelivr.net/gh/XavierJiezou/ys-dl@main/image/favicon.ico" />
<h1 align="center">原神下载器</h1>
<p align="center">一个可以从<a href="https://bbs.mihoyo.com/ys/">原神社区</a>自动抓取图片的命令行工具。
</p>
</p>
<p align="center">
    <a href="https://github.com/XavierJiezou/ys-dl/actions?query=workflow:Release">
        <img src="https://github.com/XavierJiezou/ys-dl/workflows/Release/badge.svg"
            alt="GitHub Workflow Release Status" />
    </a>
    <a href='https://ys-dl.readthedocs.io/en/latest/?badge=latest'>
        <img src='https://readthedocs.org/projects/ys-dl/badge/?version=latest' alt='Documentation Status' />
    </a>
    <a
        href="https://www.codacy.com/gh/XavierJiezou/ys-dl/dashboard?utm_source=github.com&amp;utm_medium=referral&amp;utm_content=XavierJiezou/ys-dl&amp;utm_campaign=Badge_Grade">
        <img src="https://app.codacy.com/project/badge/Grade/c2f85c8d6b8a4892b40059703f087eab" alt="Codacy Badge">
    </a>
    <a href="https://codecov.io/gh/XavierJiezou/ys-dl">
        <img src="https://codecov.io/gh/XavierJiezou/ys-dl/branch/main/graph/badge.svg?token=QpCLcUGoYx" alt="codecov">
    </a>
    <a href="https://pypi.org/project/ys-dl/">
        <img src="https://img.shields.io/pypi/pyversions/ys-dl" alt="PyPI - Python Version">
    </a>
    <a href="https://pypistats.org/packages/ys-dl">
        <img src="https://img.shields.io/pypi/dm/ys-dl" alt="PyPI - Downloads">
    </a>
    <a href="https://pypi.org/project/ys-dl/">
        <img src="https://img.shields.io/pypi/v/ys-dl" alt="PyPI">
    </a>
    <a href="https://github.com/XavierJiezou/ys-dl/stargazers">
        <img src="https://img.shields.io/github/stars/XavierJiezou/ys-dl" alt="GitHub stars">
    </a>
    <a href="https://github.com/XavierJiezou/ys-dl/network">
        <img src="https://img.shields.io/github/forks/XavierJiezou/ys-dl" alt="GitHub forks">
    </a>
    <a href="https://github.com/XavierJiezou/ys-dl/issues">
        <img src="https://img.shields.io/github/issues/XavierJiezou/ys-dl" alt="GitHub issues">
    </a>
    <a href="https://github.com/XavierJiezou/ys-dl/blob/main/LICENSE">
        <img src="https://img.shields.io/github/license/XavierJiezou/ys-dl" alt="GitHub license">
    </a>
    <br />
    <br />
    <a href="https://www.python.org/">
        <img src="http://ForTheBadge.com/images/badges/made-with-python.svg" alt="forthebadge made-with-python">
    </a>
    <a href="https://github.com/XavierJiezou">
        <img src="http://ForTheBadge.com/images/badges/built-with-love.svg" alt="ForTheBadge built-with-love">
    </a>
</p>
<p align="center">
    <a href="#演示">观看演示</a>
    •
    <a href="https://github.com/xavierjiezou/ys-dl/issues/new">报告错误</a>
    •
    <a href="https://github.com/xavierjiezou/ys-dl/issues/new">功能需求</a>
  </p>
  <p align="center">
    <a href="/docs/README.en.md">English </a>
    •
    <a href="/docs/README.cn.md">简体中文</a>
</p>
<p align="center">喜欢这个项目吗？请考虑捐赠<a href="https://paypal.me/xavierjiezou?country.x=C2&locale.x=zh_XC">（<a
            href="https://cdn.jsdelivr.net/gh/XavierJiezou/ys-dl@main/image/wechat.jpg">微信</a> | <a
            href="https://cdn.jsdelivr.net/gh/XavierJiezou/ys-dl@main/image/alipay.jpg">支付宝</a>）</a>，以帮助它改善！</p>

## 演示

![demo](https://cdn.jsdelivr.net/gh/XavierJiezou/ys-dl@main/image/demo.png)

![show](https://cdn.jsdelivr.net/gh/XavierJiezou/ys-dl@main/image/show.png)

## 功能

- [x] 原神[官网](https://ys.mihoyo.com/)图标下载
- [x] 原神[官网](https://ys.mihoyo.com/)背景图片下载
- [x] 原神社区 [COS](https://bbs.mihoyo.com/ys/home/49) 图下载
- [x] 原神社区[同人图](https://bbs.mihoyo.com/ys/home/29)下载
- [x] 原神社区 [COS 榜](https://bbs.mihoyo.com/ys/imgRanking/49/1)与[同人榜](https://bbs.mihoyo.com/ys/imgRanking/29/1)下图片载
- [ ] 原神社区[话题标签](https://bbs.mihoyo.com/ys/topicDetail/558)所属图片下载

## 安装

```bash
pip install ys-dl
```

---

`ys-dl`是基于 [selenium](https://www.selenium.dev/) 开发的。目前唯一支持的浏览器是 Chrome，因此你的系统必须安装有 Chrome 浏览器，并下载 [Chrome WebDriver](https://chromedriver.chromium.org/downloads) 放入 Python 安装目录。（想了解关于 WebDriver 的详细信息？请参阅[这里](https://www.selenium.dev/documentation/webdriver/getting_started/install_drivers/)）

## 用法

`$ ys-dl`

- 原神社区 [COS](https://bbs.mihoyo.com/ys/home/49) 图片下载

```bash
ys-dl cos
```

- 原神社区[同人图](https://bbs.mihoyo.com/ys/home/29)下载

```bash
ys-dl fan
```

- 根据排名（日榜 | 周榜 | 月榜）下载原神社区 [COS 图](https://bbs.mihoyo.com/ys/imgRanking/49/1)或[同人图](https://bbs.mihoyo.com/ys/imgRanking/29/1)

```bash
ys-dl rank
```

## 更新

更新日志见 [CHANGELOG.md](/CHANGELOG.md)

## 证书

[MIT License](/LICENSE)

## 参考

[![Readme Card](https://github-readme-stats.vercel.app/api/pin/?username=python-poetry&repo=poetry)](https://github.com/python-poetry/poetry)
[![Readme Card](https://github-readme-stats.vercel.app/api/pin/?username=SeleniumHQ&repo=selenium)](https://github.com/SeleniumHQ/selenium)
[![Readme Card](https://github-readme-stats.vercel.app/api/pin/?username=psf&repo=requests)](https://github.com/psf/requests)
[![Readme Card](https://github-readme-stats.vercel.app/api/pin/?username=Textualize&repo=rich)](https://github.com/Textualize/rich)
[![Readme Card](https://github-readme-stats.vercel.app/api/pin/?username=google&repo=python-fire)](https://github.com/google/python-fire)
[![Readme Card](https://github-readme-stats.vercel.app/api/pin/?username=pytest-dev&repo=pytest)](https://github.com/pytest-dev/pytest)


CHANGELOG.md
# Change Log

## [0.1.3] - TODO

### Fixed

- Fixed the `fake_useragent` exception: `Error occurred during loading data. Trying to use cache server https://fake-useragent.herokuapp.com`.

## [0.1.2] - 2022-05-16

### Added

- Added `rank` command.

### Changed

- Changed the `poetry.toml` file.

### Fixed

- Fixed the `cli.py` import.

## [0.1.1] - 2022-05-15

### Added

- Added `ys-dl version` command.

### Changed

- Changed `README.md` file.

### Fixed

- Fixed `version` display.

## [0.1.0] - 2022-05-14

Initial release

[0.1.3]: https://github.com/XavierJiezou/ys-dl/releases/tag/0.1.3
[0.1.2]: https://github.com/XavierJiezou/ys-dl/releases/tag/0.1.2
[0.1.1]: https://github.com/XavierJiezou/ys-dl/releases/tag/0.1.1
[0.1.0]: https://github.com/XavierJiezou/ys-dl/releases/tag/0.1.0

tests/test_ys_dl.py
import os
import random
import time

from selenium import webdriver
import toml

from ys_dl.core import YsDl
from ys_dl.apis import APIS


class TestYsDl(YsDl):

    options = webdriver.ChromeOptions()
    options.add_experimental_option('excludeSwitches', ['enable-logging'])

    def test_version(self):
        project = toml.load('pyproject.toml')
        version = project['tool']['poetry']['version']
        assert self.version == version

    def test_dl_img_static(self):
        sample = {
            'https://upload-bbs.mihoyo.com/upload/2022/05/01/301685713/8f168d7cd4aef6c7f556d882712fc379_1319607209990236918.jpg': 3234434,
            'https://upload-bbs.mihoyo.com/upload/2022/05/04/295576484/2f9846ab3f86700ec6d9ebf1da034f48_2276417862705020206.jpg': 7658847,
            'https://upload-bbs.mihoyo.com/upload/2022/05/09/214325571/ce6a12088bbbfa629b4d5fc46f31f271_2525566984544398307.jpg': 4003035,
        }
        for url in sample:
            name = os.path.basename(url)
            self.dl_img_static(url, name)
            size = os.path.getsize(name)
            assert size == sample[url]
            time.sleep(1)

    def test_dl_img_dynamic(self):
        sample = {
            'https://upload-bbs.mihoyo.com/upload/2022/05/01/301685713/8f168d7cd4aef6c7f556d882712fc379_1319607209990236918.jpg': 3234434,
            'https://upload-bbs.mihoyo.com/upload/2022/05/04/295576484/2f9846ab3f86700ec6d9ebf1da034f48_2276417862705020206.jpg': 7658847,
            'https://upload-bbs.mihoyo.com/upload/2022/05/09/214325571/ce6a12088bbbfa629b4d5fc46f31f271_2525566984544398307.jpg': 4003035,
        }
        driver = webdriver.Chrome(options=self.options)
        for url in sample:
            save_dir = os.getcwd()
            img_name = os.path.basename(url)
            save_path = os.path.join(save_dir, img_name)
            if os.path.exists(save_path):
                pass
            else:
                self.dl_img_dynamic(url, save_path, driver)
                size = os.path.getsize(save_path)
                assert size == sample[url]
        driver.quit()

    def test_dl_img(self):
        sample = {
            'https://upload-bbs.mihoyo.com/upload/2022/05/01/301685713/8f168d7cd4aef6c7f556d882712fc379_1319607209990236918.jpg': 3234434,
            'https://upload-bbs.mihoyo.com/upload/2022/05/04/295576484/2f9846ab3f86700ec6d9ebf1da034f48_2276417862705020206.jpg': 7658847,
            'https://upload-bbs.mihoyo.com/upload/2022/05/09/214325571/ce6a12088bbbfa629b4d5fc46f31f271_2525566984544398307.jpg': 4003035,
        }
        driver = webdriver.Chrome(options=self.options)
        for url in sample:
            save_dir = os.getcwd()
            img_name = os.path.basename(url)
            save_path = os.path.join(save_dir, img_name)
            if os.path.exists(save_path):
                pass
            else:
                mode = random.choice(['static', 'dynamic'])
                self.dl_img(url, mode, save_dir, img_name, driver)
                size = os.path.getsize(save_path)
                assert size == sample[url]
        driver.quit()

    def test_get_icon_img(self):
        result = self.get_icon_img()
        icon_url = result['data']
        assert '.ico' in icon_url

    def test_get_bg_img(self):
        result = self.get_bg_img()
        bg_url = result['data']
        assert '.jpg' in bg_url

    def test_get_article_url(self):
        sample = {
            'cos-rank-daily': 1,
            'cos-rank-daily': 20,
            'cos-rank-daily': 100,
            'cos-hot': 1,
            'cos-hot': 20,
            'cos-hot': 100,
            'cos-latest_post': 1,
            'cos-latest_post': 20,
            'cos-latest_post': 100,
            'cos-latest_reply': 1,
            'cos-latest_reply': 20,
            'cos-latest_reply': 100,
            'fan-rank-daily': 1,
            'fan-rank-daily': 100,
            'fan-rank-daily': 200,
            'fan-hot': 1,
            'fan-hot': 20,
            'fan-hot': 100,
            'fan-latest_post': 1,
            'fan-latest_post': 20,
            'fan-latest_post': 100,
            'fan-latest_reply': 1,
            'fan-latest_reply': 20,
            'fan-latest_reply': 100,
        }
        self.options.add_argument('--headless')
        driver = webdriver.Chrome(options=self.options)
        for key in sample:
            result = self.get_article_url(
                driver,
                *key.split('-'),
                sample[key]
            )
            if 'rank' not in key:
                tmp_var = sample[key]
            else:
                if 'cos' in key:
                    limit_length = 20
                else:
                    limit_length = 100
                tmp_var = min(sample[key], limit_length)
            assert tmp_var == len(result['data'])
        driver.quit()

    def test_get_article_img(self):
        sample = {
            'https://bbs.mihoyo.com/ys/article/21477272': 9,
            'https://bbs.mihoyo.com/ys/article/21735163': 4,
            'https://bbs.mihoyo.com/ys/article/22054203': 3,
            'https://bbs.mihoyo.com/ys/article/21398220': 1,
            'https://bbs.mihoyo.com/ys/article/22255184': 5,
            'https://bbs.mihoyo.com/ys/article/22142359': 2,
        }
        self.options.add_argument('--headless')
        driver = webdriver.Chrome(options=self.options)
        for url in sample:
            result = self.get_article_img(url, driver)
            assert len(result['data']['urls']) == sample[url]
        driver.quit()
    
    def test_get_article(self):
        pass

    def test_get_topic_url(self):
        pass

    def test_get_topic_img(self):
        pass
    
    def test_get_topic(self):
        pass


tests/test_ys_dl_cli.py
import toml

from ys_dl.cli import YsDlCli


class TestYsDlCli(object):

    ysdlcli = YsDlCli()

    def test_init(self):
        try:
            self.ysdlcli
        except:
            raise

    def test_version(self):
        project = toml.load('pyproject.toml')
        version = project['tool']['poetry']['version']
        assert self.ysdlcli.version == version

    def test_icon(self):
        pass

    def test_bg(self):
        pass

    def test_get_article_label(self):
        pass

    def test_get_rank_key(self):
        pass

    def test_get_rank_api(self):
        pass

    def test_get_dl_limit(self):
        pass

    def test_dl_img_loop(self):
        pass

    def test_get_params(self):
        pass

    def test_article(self):
        pass

    def test_cos(self):
        pass

    def test_fan(self):
        pass

    def test_rank(self):
        pass

    def test_topic(self):
        pass


ys_dl/cli.py
import json
import os
import re
import time
from typing import Any

import fire
from rich.console import Console
from rich.progress import (
    Progress,
    SpinnerColumn,
    MofNCompleteColumn,
    TimeElapsedColumn,
)
from rich.prompt import Prompt, IntPrompt
from rich.theme import Theme
from selenium import webdriver

from ys_dl.core import YsDl
from ys_dl.apis import APIS


class YsDlCli(object):

    def __init__(self) -> None:
        self.options = webdriver.ChromeOptions()
        self.options.add_experimental_option(
            'excludeSwitches',
            ['enable-logging']
        )
        theme = Theme({
            "success": "cyan",
            "failure": "b red",
        })
        self.console = Console(theme=theme)
        self.progress = Progress(
            SpinnerColumn(),
            MofNCompleteColumn(),
            *Progress.get_default_columns(),
            TimeElapsedColumn(),
            console=self.console,
        )

    @property
    def version(self) -> str:
        return YsDl().version

    def icon(self) -> None:
        size = Prompt.ask(
            prompt='Select the size',
            choices=['256', '32'],
            default='256'
        )
        result = YsDl().get_icon_img(size=int(size))
        if result['status'] == 200:
            with self.console.status(f'Downloading {size}x{size} icon to local'):
                YsDl().dl_img(result['data'])
        else:
            pass
        self.console.print_json(json.dumps(result))

    def bg(self) -> None:
        result = YsDl().get_bg_img()
        if result['status'] == 200:
            with self.console.status('Downloading background image to local'):
                YsDl().dl_img(result['data'], img_name='bg.jpg')
        else:
            pass
        self.console.print_json(json.dumps(result))

    def _get_article_label(self) -> str:
        labels = ['rank', 'hot', 'latest_post', 'latest_reply']
        emojis = [
            ':rocket:',
            ':fire:',
            ':envelope_with_arrow:',
            ':loudspeaker:',
        ]
        tmp_str = '/'.join([f'{e} {l}' for e, l in zip(emojis, labels)])
        return Prompt.ask(
            prompt=f'Enter the label [b][magenta][{tmp_str}][/magenta][/b]',
            choices=labels,
            show_choices=False,
            default='rank',
        )

    def _get_rank_key(self) -> str:
        return Prompt.ask(
            prompt='Select the key',
            choices=['daily', 'weekly', 'monthly'],
            default='daily',
        )

    def _get_rank_api(self) -> str:
        return Prompt.ask(
            prompt='Select the api',
            choices=['cos', 'fan'],
            default='cos',
        )

    def _get_dl_limit(self, max_value: int, default_value: int) -> int:
        while True:
            limit = IntPrompt.ask(
                prompt=f'Enter the number of pages to download [b][magenta][0~{max_value}][/magenta][/b]',
                default=default_value,
            )
            if 0 <= limit <= max_value:
                break
            self.console.print(
                f':pile_of_poo: [prompt.invalid]Number must be between 0 and {max_value}')
        return limit

    def _dl_img_loop(
        self,
        article_urls: list[str],
        driver: webdriver,
        is_rank: bool,
        params: list[Any],
    ) -> None:
        if is_rank:
            overall_desc = f'{params[1]}/{params[0]}/{params[2]}'
        else:
            if params[1] == 'rank':
                overall_desc = '/'.join(params[:3])
            else:
                overall_desc = '/'.join(params[:2])
        with self.progress:
            task = self.progress.add_task(
                description=f'[green]{overall_desc}',
                total=len(article_urls),
            )
            count = 0
            for idx, url in enumerate(article_urls):
                image = YsDl().get_article_img(url, driver)
                title = image['data']['title'].replace(' ', '')
                title = re.sub('[/\\:*?"<>|]', '', title)
                urls = image['data']['urls']
                save_dir = os.path.join(
                    f'{overall_desc}/',
                    f'[{idx+1:0>{len(str(len(article_urls)))}}]{title}/',
                )
                # if len(title)<=10:
                #     _title = title
                # else:
                #     _title = f'{title[:7]}...'
                sub_task = self.progress.add_task(
                    description=f'[cyan][{idx+1:0>{len(str(len(article_urls)))}}]{title}',
                    total=len(urls),
                )
                for _idx, _url in enumerate(urls):
                    YsDl().dl_img(
                        img_url=_url,
                        save_dir=save_dir,
                        img_name=f'{_idx+1}.jpg',
                    )
                    save_path = os.path.join(save_dir, f'{_idx+1}.jpg')
                    link_path = os.path.join(os.getcwd(), save_dir, f'{_idx+1}.jpg')
                    self.console.log(f'Downloaded {save_path}', style=f'link {link_path}')
                    self.progress.update(sub_task, advance=1)
                    time.sleep(1)
                    count += 1
                self.progress.update(task, advance=1)
        with self.console.status('Quitting the webdriver'):
            driver.quit()
        self.console.print(f'[green]{count} images downloaded, done!')

    def _get_params(self, api_key: str, is_rank: bool) -> list[Any]:
        if is_rank:
            sub_key = 'rank'
            rank_key = self._get_rank_key()
            max_value = APIS[api_key][sub_key]['limit_length']
            url_nums = self._get_dl_limit(max_value, max_value)
        else:
            label = self._get_article_label()
            if label == 'rank':
                sub_key = 'rank'
                rank_key = self._get_rank_key()
                max_value = APIS[api_key][sub_key]['limit_length']
                url_nums = self._get_dl_limit(max_value, max_value)
            else:
                sub_key = label
                rank_key = None
                max_value = 500
                url_nums = self._get_dl_limit(max_value, 100)
        return [
            api_key,
            sub_key,
            rank_key,
            url_nums,
        ]

    def _article(self, api_key: str = 'cos', is_rank: bool = False) -> None:
        with self.console.status('Yuanshen downloader booting up'):
            time.sleep(1)
        with self.console.status('Creating a webdriver (headless)'):
            self.options.add_argument('--headless')
            driver = webdriver.Chrome(options=self.options)
        params = self._get_params(api_key, is_rank)
        with self.console.status('Getting urls to articles'):
            result = YsDl().get_article_url(
                driver,
                *params,
            )
        self._dl_img_loop(result['data'], driver, is_rank, params)

    def cos(self) -> None:
        self._article('cos')

    def fan(self) -> None:
        self._article('fan')

    def rank(self) -> None:
        api_key = self._get_rank_api()
        self._article(api_key, is_rank=True)

    # TODO (GitHub@XavierJiezou): Download images belonging to topic tags.
    def topic(self) -> str:
        return 'TODO (GitHub@XavierJiezou): Download images belonging to topic tags.'


def main():
    fire.Fire(YsDlCli)


if __name__ == '__main__':
    main()


ys_dl/core.py
import os
import re
import time
from typing import Any

from fake_useragent import UserAgent
import pyautogui
import pyperclip
import requests
from selenium import webdriver
from selenium.webdriver.common.by import By
from selenium.webdriver.common.keys import Keys
from selenium.webdriver.common.action_chains import ActionChains

from ys_dl.apis import APIS
from ys_dl import __version__


class YsDl(object):

    @property
    def version(self):
        return __version__

    def dl_img_static(
        self,
        img_url: str,
        save_path: str,
    ) -> dict[str, Any]:
        headers = {'user-agent': UserAgent().random}
        res = requests.get(img_url, headers=headers)
        if res.status_code == 200:
            with open(save_path, 'wb') as f:
                f.write(res.content)
            return {
                'status': res.status_code,
                'data': save_path,
                'msg': 'success',
            }
        else:
            return {
                'status': res.status_code,
                'data': None,
                'msg': 'failure',
            }

    def dl_img_dynamic(
        self,
        img_url: str,
        save_path: str,
        driver: webdriver,
    ) -> dict[str, Any]:
        driver.get(img_url)
        driver.implicitly_wait(5)
        pyautogui.hotkey('ctrlleft', 's')
        time.sleep(1)
        pyperclip.copy(save_path)
        time.sleep(0.5)
        pyautogui.hotkey('ctrlleft', 'v')
        time.sleep(0.5)
        pyautogui.press('enter')
        time.sleep(1)
        if os.path.exists(save_path):
            return {
                'status': 200,
                'data': save_path,
                'msg': 'success',
            }
        else:
            return {
                'status': 404,
                'data': None,
                'msg': 'failure',
            }

    def dl_img(
        self,
        img_url: str,
        mode: str = 'static',
        save_dir: str = None,
        img_name: str = None,
        driver: webdriver = None,
    ) -> dict[str, Any]:
        if not save_dir:
            save_dir = os.getcwd()
        else:
            os.makedirs(save_dir, exist_ok=True)
        if not img_name:
            img_name = os.path.basename(img_url)
        else:
            pass
        save_path = os.path.join(save_dir, img_name)
        if os.path.exists(save_path):
            return {
                'status': 200,
                'data': save_path,
                'msg': 'success',
            }
        else:
            if mode == 'static':
                return self.dl_img_static(img_url, save_path)
            else:
                return self.dl_img_dynamic(img_url, save_path, driver)

    def get_icon_img(
        self,
        api_key: str = 'icon',
        size: int = 256,
    ) -> dict[str, Any]:
        url = APIS.get(api_key).get(size)
        return {
            'status': 200,
            'data': url,
            'msg': 'success'
        }

    def get_bg_img(self, api_key: str = 'bg') -> dict[str, Any]:
        url = APIS.get(api_key)
        headers = {'user-agent': UserAgent().random}
        res = requests.get(url=url, headers=headers)
        if res.status_code == 200:
            pattern = re.compile('https://web.*?config_.*?.js')
            js_url = re.findall(pattern, res.text)[0]
            js_res = requests.get(url=js_url, headers=headers)
            if js_res.status_code == 200:
                pattern = re.compile('resources":{"(.*?.jpg)"')
                bg_url = re.findall(pattern, js_res.text)[0]
                return {
                    'status': js_res.status_code,
                    'data': bg_url,
                    'msg': 'success'
                }
            else:
                return {
                    'status': js_res.status_code,
                    'data': None,
                    'msg': 'failure'
                }
        else:
            return {
                'status': res.status_code,
                'data': None,
                'msg': 'failure'
            }

    def get_article_url(
        self,
        driver: webdriver,
        api_key: str = 'cos',
        sub_key: str = 'rank',
        rank_key: str = 'daily',
        url_nums: int = 100,
        max_trying: int = 100,
    ) -> dict[str, Any]:
        url = APIS[api_key][sub_key]
        if sub_key == 'rank':
            url_nums = min(url['limit_length'], url_nums)
            url = url[rank_key]
        else:
            pass
        driver.maximize_window()
        driver.get(url)
        driver.implicitly_wait(5)
        action = ActionChains(driver)
        count = 0
        for _ in range(max_trying):
            tag_a = driver.find_elements(
                By.CSS_SELECTOR,
                APIS['css_selector']['article']['url'],
            )
            if len(tag_a) >= url_nums:
                break
            else:
                action.send_keys(Keys.PAGE_DOWN).perform()
                time.sleep(0.5)
                count += 1
        if count < max_trying:
            urls = []
            for _ in tag_a[:url_nums]:
                urls.append(_.get_attribute('href'))
            return {
                'status': 200,
                'data': urls,
                'msg': 'success',
            }
        else:
            return {
                'status': 404,
                'data': None,
                'msg': 'failure'
            }

    def get_article_img(
        self,
        article_url: str,
        driver: webdriver,
        size: str = 'large'
    ) -> dict[str, Any]:
        driver.maximize_window()
        driver.get(article_url)
        driver.implicitly_wait(5)
        value = APIS['css_selector']['article']['title']
        title = driver.find_element(By.CSS_SELECTOR, value).text
        value = APIS['css_selector']['article']['image']
        image = driver.find_elements(By.CSS_SELECTOR, value)
        if image:
            if size == 'large':
                urls = [_.get_attribute('large') for _ in image]
            else:
                urls = [_.get_attribute('src') for _ in image]
            return {
                'status': 200,
                'data': {
                    'title': title,
                    'urls': urls,
                },
                'msg': 'success'
            }
        else:
            return {
                'status': 404,
                'data': None,
                'msg': 'failure'
            }
    
    def get_article(self):
        pass

    # TODO (GitHub@XavierJiezou): Download images belonging to topic tags.
    def get_topic_url(self):
        pass

    # TODO (GitHub@XavierJiezou): Download images belonging to topic tags.
    def get_topic_img(self):
        pass

    # TODO (GitHub@XavierJiezou): Download images belonging to topic tags.
    def get_topic(self):
        pass


if __name__ == '__main__':
    options = webdriver.ChromeOptions()
    options.add_experimental_option('excludeSwitches', ['enable-logging'])
    driver = webdriver.Chrome(options=options)
    result = YsDl().get_article_img('https://bbs.mihoyo.com/ys/article/22335059', driver=driver)
    print(result)
    driver.quit()

ys_dl/apis.py
"""A summary of interesting image APIs to https://bbs.mihoyo.com/ys/"""


from ast import Pass


APIS = {
    'icon': {
        32: 'https://ys.mihoyo.com/main/favicon.ico',
        256: 'https://genshin.mihoyo.com/favicon.ico',
    },
    'bg': 'https://ys.mihoyo.com/',
    'cos': {

        'hot': 'https://bbs.mihoyo.com/ys/home/49?type=hot',
        'latest_reply': 'https://bbs.mihoyo.com/ys/home/49?type=1',
        'latest_post': 'https://bbs.mihoyo.com/ys/home/49?type=2',
        'rank': {
            'daily': 'https://bbs.mihoyo.com/ys/imgRanking/49/1',
            'weekly': 'https://bbs.mihoyo.com/ys/imgRanking/49/2',
            'monthly': 'https://bbs.mihoyo.com/ys/imgRanking/49/3',
            'limit_length': 20
        },
    },
    'fan': {
        'hot': 'https://bbs.mihoyo.com/ys/home/29?type=hot',
        'latest_reply': 'https://bbs.mihoyo.com/ys/home/29?type=1',
        'latest_post': 'https://bbs.mihoyo.com/ys/home/29?type=2',
        'rank': {
            'daily': 'https://bbs.mihoyo.com/ys/imgRanking/29/1',
            'weekly': 'https://bbs.mihoyo.com/ys/imgRanking/29/2',
            'monthly': 'https://bbs.mihoyo.com/ys/imgRanking/29/3',
            'limit_length': 100
        },
    },
    'topic': { # TODO (GitHub@XavierJiezou): Download images belonging to topic tags.
        'emoji': {
            'new': 'https://bbs.mihoyo.com/ys/topicDetail/105',
            'select': 'https://bbs.mihoyo.com/ys/topicDetail/105?type=1',
            'hot': 'https://bbs.mihoyo.com/ys/topicDetail/105?type=2',
        },
        'wallpaper': {
            'new': 'https://bbs.mihoyo.com/ys/topicDetail/558',
            'hot': 'https://bbs.mihoyo.com/ys/topicDetail/558?type=2',
        },
    },
    'css_selector': {
        'article': {
            'url': 'div.mhy-article-list__body>div>a.mhy-router-link',
            'title': 'div.mhy-article-page__title>h1',
            'image': 'div.mhy-img-article img',
        },
        'topicDetail': {
            'url': '',
            'title': '',
            'image': '',
        },
    },
}


if __name__ == '__main__':
    pass
